### DISCUSSION TITLE
Prepare data for texgen webui fine tuning

### DISCUSSION THREAD

**Discussion by EnesK55** [oobabooga/text-generation-webui#3038] [Q&A]
Hello everyone, 

sorry if this is the wrong category to place this topic. 

First of all, I have no experience in this area and I just started to work on language models. 
Currently I am trying to create a language model and fine tune it with my data. 
My data is generated from scientific papers (i.e. from science direct). 
My question is: What should my data look like and which tools do you use to extract data from scietific paper and how to clean the data?

Appreciate any answer, thank you in advance!

**Comment by FartyPants** [comment]
But you don't want that  with Lora/Qlora. I hope not, for your own sake and the future of the scientific community. 

Lora's purpose is to add just a small % (0.2-3%) of parameters to the model and keep all previous parameters frozen.
You will manage to make an adapter that will adopt the language of your scientific papers, even happily use some often repeating themes (names, words, ideas) and totally BS you with the actual details. Your eyes would roll. 
Beyond being entertained by it or impressing clueless people on youtube, there is no other value. 
Your biggest issue is that none of these models are capable of even remotely sensing they don't know something. The less they know, the more BS will come out. 

You want to look at full fine-tuning at least - check https://github.com/OpenAccess-AI-Collective/axolotl
That also requires runpod or a machine with a lot of VRAM 
Or you want to look at langchain/indexing where the model looks at the actual papers and quote/rephrase from them. That's something you can do on a shoestring budget.

See I just saved you a week of frustration with Lora. Wrap the bottle in bubble paper so it doesn't break in the mail.

**Reply by EnesK55** [reply]
We should have that wine when we meet, you seem like a fun, cool guy (based on your reply) :D Actually I was already depressed today, because I tried a lot of models by using textgen webui and it did not work properly, but I thought my dataset is probably too small (500k words). Additionally, since I extract data from scientific papers, it is not really "clean". The txt file includes a lot of useless information (like mail adress, websites etc.), therefore my plan was to clean the data and also increase the amount of words and give it another try....

Max 3% sounds horrible to be honest, can not really understand the huge hype if that is the case!
But I found a quite interesting story about training an alpaca model by using an unreal engine 5 documentation. The results were quite good in my opinion, therefore I thought it could work for my case too. 
I am using a RTX A5000 not sure if this is good or bad, but I considered to train/finetune a llama 7B from scratch (I guess you will start laughing here). 
I already tried privatechatGPT which wasnt that bad but a little slow. 

Thank you so much for your reply, I really appreciate your suggestions and thoughts!!!  I'll definitely check out the github link you sent.

**Comment by EnesK55** [comment]
First of all thank you very much for your great explanations, I really
appreciate it!!

   - I totally understood your point and it makes sense even if I do not
   have much knowledge! Since I have to try this method at least for an
   experimental purpose, which model you would recommend to me?
   My purpose is to create an AI expert in terms of printed circuit board
   soldering.
   Is it wrong to say, that if I have a smaller dataset, it is better to
   use a smaller model with higher rank, like 7B? Since you mentioned that a
   small rank would only affect by 3%, I would always go for higher rank.
   - Could you recommend any books or something else to gain knowledge in
   terms of language models
   - Furthermore, I noticed that when I use "privateGPT" (where you can
   "talk" to your pdf) the answers are really good, despite of my unfiltered
   data. Is that a sign that my data is not that bad?
   - Last question: Not every model is supported by lora training, is there
   another way to use texgen webui to fine tune a modell like galactica?

What is the common method to extract text from scientific  pdf papers?

Am Sa., 8. Juli 2023 um 11:20 Uhr schrieb FartyPants <
***@***.***>:

**Comment by FartyPants** [comment]
No matter what, if you attempt to fine-tune with LoRA any model on ANY NEW data, the result will be that it will speak well in the terms of your document, with a high chance of making stuff up. It's fine for chatting, but not very good if you actually want answers.

The privateGPT is indeed the proper way and there will be or maybe already are extensions in ooba that do the same. Because everyine will soon realize this issue, that LoRA loves to BS even more, hahaha
The idea is: get the factual data from document and shove it to LLM at the time of interference.

Here is a bonus idea:
- given the premise that such privateGPT extension exists or will exist in ooba soon
- train LoRA on the document
- use the model+loRA + that privateGPT extension and it should give you far more accurate response as it would be biased to answer in the document way

To some of the answers:
using 7b vs 13b, you could have a bit higher chance to find sweet spot where the document is significant yet not overfitting the model - but then 7b models are less logical - so hard to say really... YES and NO,  this is more voodoo than a science (coming from a man who trained about 100 LoRA's and still didn't figure the right values - scratch that 175 LoRA's as of now)

as for model, I 99% use Wizard-Vicuna-13b-Uncensored for my further experiments. Guanaco seems to be well regarded too but I have weird EOS issues on whatever guanaco I downloaded and was too lazy to get proper one.
Don't mess with the "SuperHot" variants - it's a pretend-play for children so they feel like they can play with the same tools like big boys at OpenAI. No they can't. Not even close.

**Comment by EnesK55** [comment]
Thank you very much that you are helping me, I appreciate every single word!!

The idea of combining model, lora and privateGPT extension sounds really good, but I am far away from that at the moment...
Since I am not even sure how to use Textgen webui properly.
Do I have to modify something after loading the model?
Because after "Successfully loaded TheBloke_Wizard-Vicuna-13B-Uncensored-HF" I go to text generation and ask my questions, but I get no response.
Furthermore, how can I check if I am really talking to the vicuna model, is there a specific question like "To which model I am talking to?"?

Question that bothers me the most: 
How has my data to look like. 
When I search for "Text Cleaning Methods in NLP", I get results like: remove punctuation, numbers, lowcasing the data...
I am not sure if this is still needed for the modern llm´s?
The unreal engine 5 documentation did not include the cleaning methods and it worked, but I read in his repo, that he would recommend to use a better format for the data, but how would a better format look like. Because I could not find any dataset which is not question/answer pairs.
So, what is the proper text data format for the lora training beside question/answer pairs!?
